Revise doc comments for structured concurrency

amartini51 opened the pull request:
Technical and non-clean-up corrections:

- Removed the mention of "if no current task is available" from
  `Task.priority` because it's an instance variable.  By definition, you
  can't access it without some current task.
- Changed the abstract for `detach(priority:operation:)` to note that
  it's for non-throwing operations, since it uses `Never` as its
  `Task.Handle` failure type.
- Removed discussion of throwing and rethrowing errors from `TaskGroup`,
  since its child tasks can't be throwing.
- Removed explicit see-also cross references to a type that's being used
  in the declaration.  The rendered documentation already provides an
  affordance to navigate to a type in that context.

Style changes, per Apple Style Guide and reference style guidelines:

- Used "canceled" and "canceling" throughout.  (But note "cancellation".)
- Used contractions.
- Removed third-level headings.  The discussion sections they were in
  are short enough that they doesn't need headings.
- Removed code voice from abstracts.
- Used noun phrase abstracts (not "Returns ...") for properties, instead
  of describing them as if they were methods.
- Referred to methods by their full names, not just the part before the
  parentheses.  For example, `spawn(overridingPriority:operation:)`
  rather than just `spawn` or `spawn()`.

Other changes:

- Marked headings in type overviews by underling with a line of equal
  signs (=), to match markup elsewhere in the standard library.
- Included the type names when referring to methods, to avoid ambiguity
  and to make it possible to look up the documentation for that method
  if the reader doesn't already know what type it's on.

For <rdar://76086293>

amartini51 committed:
Clean up doc comments for task APIs.

Technical and non-clean-up corrections:

- Removed the mention of "if no current task is available" from
  `Task.priority` because it's an instance variable.  By definition, you
  can't access it without some current task.
- Changed the abstract for `detach(priority:operation:)` to note that
  it's for non-throwing operations, since it uses `Never` as its
  `Task.Handle` failure type.
- Removed discussion of throwing and rethrowing errors from `TaskGroup`,
  since its child tasks can't be throwing.
- Removed explicit see-also cross references to a type that's being used
  in the declaration.  The rendered documentation already provides an
  affordance to navigate to a type in that context.

Style changes, per Apple Style Guide and reference style guidelines:

- Used "canceled" and "canceling" throughout.  (But note "cancellation".)
- Used contractions.
- Removed third-level headings.  The discussion sections they were in
  are short enough that they doesn't need headings.
- Removed code voice from abstracts.
- Used noun phrase abstracts (not "Returns ...") for properties, instead
  of describing them as if they were methods.
- Referred to methods by their full names, not just the part before the
  parentheses.  For example, `spawn(overridingPriority:operation:)`
  rather than just `spawn` or `spawn()`.

Other changes:

- Marked headings in type overviews by underling with a line of equal
  signs (=), to match markup elsewhere in the standard library.
- Included the type names when referring to methods, to avoid ambiguity
  and to make it possible to look up the documentation for that method
  if the reader doesn't already know what type it's on.

amartini51 commented:
Marking as a draft until the placeholders (marked with ◊) get filled in.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -15,22 +15,39 @@ import Swift
+/// by invoking the appropriate context-sensitive static functions which operate
+/// on the current task. Because all such functions are asynchronous, they can only
 /// be invoked as part of an existing task, and therefore are guaranteed to be
 /// effective.
```
Sorry, this is outdated wording here; 

We have both static (non async) functions and instance functions now. The wording about async functions and properties is pretty ancient sorry.

I think we can remove the entire sentence from "Because all such functions are asynchronous ... " since those dont exist anymore and they're handled by the static and instance ones.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -15,22 +15,39 @@ import Swift
+/// on the current task. Because all such functions are asynchronous, they can only
 /// be invoked as part of an existing task, and therefore are guaranteed to be
 /// effective.
+/// ◊TR: What do we mean by "effective" above?
```
I think we can remove the entire sentence so no need to dwell here; but what I meant that "they will work and not crash utterly" :)

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -15,22 +15,39 @@ import Swift
-/// Partial tasks are generally not interacted with by end-users directly,
-/// unless implementing a scheduler.
+/// Unless you're implementing a scheduler,
+/// you don't generally interacted with partial tasks directly.
```
```suggestion
/// you don't generally interact with jobs directly.
```

I believe John is removing the notion of partial tasks and they'll be named "Jobs": https://github.com/apple/swift/pull/36878/files#diff-92484f8fd5d76f2bd84b8509dc599fde13def3c55f342c8a64ffbba2cc9d7c97L7

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -15,22 +15,39 @@ import Swift
+///
+/// ◊ a task checks its own cancellation status
+/// ◊ you might need to check cancellation repeatedly for long-running work
+/// ◊ the canonical response to cancelation is to throw `Task.CancellationError`
```
Yes, all these are correct.

The shortcut to "check and throw the cancellation error" is `try Task.checkCancellation()` but one might implement it just with `Task.isCancelled` and throwing

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -46,10 +63,9 @@ public struct Task {
 
-  /// Returns 'current' `Task` instance, representing the task from within which
-  /// this function was called.
+  /// Returns the task that this code is being run on.
```
I feel we should expand here, how about something like:

```suggestion
  /// Returns the task that this code is being run on.
  /// 
  /// If invoked from the context of an asynchronous function or closure, 
  /// this is guaranteed to return a non-nil task. When called from a synchronous
  /// context the returned value depends on wether the synchronous operation was
  /// itself called from an asynchronous context or not. For example:
  /// 
  ///    func hello() { assert(Task.current != nil) }
  ///    func asynchronous() async { hello() } 
  ///
  /// is correct and will not trigger the assertion, because an asynchronous function exists in the call-stack leading to the `Task.current` invocation.
```

or something like that?

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -70,58 +86,58 @@ extension Task {
   ///
-  /// If no current `Task` is available, returns `Priority.default`.
+  /// If you access this property outside of any task,
+  /// its value is `Priority.default`.
```
FIXME: I thought we fixed this -- this should be `.unspecified` (and in implementation as well)

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -70,58 +86,58 @@ extension Task {
-  /// TODO: Define the details of task priority; It is likely to be a concept
+  /// ◊TODO: Define the details of task priority; It is likely to be a concept
   ///       similar to Darwin Dispatch's QoS; bearing in mind that priority is not as
   ///       much of a thing on other platforms (i.e. server side Linux systems).
```
for better or worse nothing seems to have been done here and we're likely to stick to the ones we have -- the todo will be likely just removed.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -70,58 +86,58 @@ extension Task {
-  ///   - this DOES affect `Task.currentPriority()`.
+  /// In both cases, priority elevation prevents a low-priority task
+  /// blocking the execution of a high priority task,
+  /// also known as *priority inversion*.
```
Hmm... I'm not sure about "prevents" it definitely "attempts to lessen the impact of" but we don't really fully prevent it... we still need to complete running the low priority task, even though now it became boosted to high priority... I guess maybe that counts as "prevents"? a bit iffy on the wording hm. @rjmccall maybe has a stronger opinion?

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -143,57 +159,58 @@ extension Task {
-  /// i.e. the task will run regardless of the handle still being present or not.
-  /// Dropping a handle however means losing the ability to await on the task's result
-  /// and losing the ability to cancel it.
+  /// It's not a programming error to discard a task's handle without awaiting or canceling the task.
```
Small side question, just to educate myself; so guidelines are to actually use "it's"? All my life in technical writing i've been avoiding it :) Is it because it's shorter/easier to read?

amartini51 replied:
I'm not sure the exact reason, but yes that's the style used here.  For documentation, we follow both [Apple Style Guide](https://help.apple.com/applestyleguide/) and style guidance from DevPubs that's specific to reference.  ASG's entry for *contractions* begins:

> As part of Apple’s informal voice, contractions are used and recommended throughout most documentation, interface text, and marketing copy. Keep localization in mind when deciding how or whether to use contractions.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -143,57 +159,58 @@ extension Task {
+  /// It's not a programming error to discard a task's handle without awaiting or canceling the task.
+  /// A task runs whether or not you still have its handle stored somewhere.
+  /// However, if you discard a task's handle, you give up the ability
+  /// to wait for that task's result or cancel the task.
```
👍

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -143,57 +159,58 @@ extension Task {
-    /// ### Cancellation
-    /// If the awaited on task gets cancelled externally the `get()` will throw
-    /// a cancellation error.
+    /// If task is canceled, this method throws a cancellation error.
```
Worth spelling out which error?

```suggestion
    /// If task is canceled, this method throws a `Task.CancellationError`.
```

?

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -143,57 +159,58 @@ extension Task {
-    /// ### Cancellation
-    /// If the awaited on task gets cancelled externally the `get()` will throw
-    /// a cancellation error.
+    /// If task is canceled, this method throws a cancellation error.
```
we lost the "externally" here:

```suggestion
    /// If task is canceled externally, this method throws a cancellation error.
```

Although I'll double check if we actually do so... I don't think we do to be honest 🤔

ktoso replied:
Are we sure we want those semantics...? I don't think we actually do this today; it'd require "check if the task we refer to was cancelled and throw magically here" I'm not sure this is actually easy to know vs. the "internally"... 😖 

@DougGregor @jckarter

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -143,57 +159,58 @@ extension Task {
+    /// If the task is canceled internally ---
+    /// for example, by checking for cancellation
+    /// and throwing a specific error, or by using `checkCancellation` ---
+    /// this method rethrows the error that was thrown by the task.
```
This wording is slightly off and ancient; most recent source of truth is https://github.com/apple/swift-evolution/blob/main/proposals/0304-structured-concurrency.md

I think the wording is rather:

```swift
/// If the task was cancelled it may check for it and throw a `Task.CancellationError`.
/// It may also decide to handle cancellation in other ways, including returning a best-effort result or similar.
```

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -143,57 +159,58 @@ extension Task {
+    /// If the task is canceled internally ---
+    /// for example, by checking for cancellation
+    /// and throwing a specific error, or by using `checkCancellation` ---
+    /// this method rethrows the error that was thrown by the task.
```
same notes as above; this wording is wrong nowadays I believe

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -221,17 +238,15 @@ extension Task.Handle where Failure == Never {
+  /// The task that this handle refers to might check for cancellation ---
+  /// however, since this method is nonthrowing,
+  /// the task can't throw `CancellationError` and needs to use another method
+  /// like returning `nil` instead.
```
sounds good

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -277,6 +292,8 @@ extension Task {
   ///
   /// This is a port of the C++ FlagSet.
+  /// ◊TR: What is FlagSet?
+  /// ◊TR: Why does that matter to users that it matches something in C++?
```
This is an internal type, so we don't need to document it in depth for end users

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -375,38 +392,26 @@ extension Task {
+/// Avoid using a detached task unless it isn't possible
+/// to model the operation using structured concurrency features like child tasks.
+/// Child tasks inherit the parent task's priority,
+/// task-local storage, and deadlines,
```
sadly we're not shipping deadlines this year

```suggestion
/// task-local storage, 
```

they'll come in the future though :)

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -375,38 +392,26 @@ extension Task {
+/// Child tasks inherit the parent task's priority,
+/// task-local storage, and deadlines,
+/// and canceling a parent task automatically cancels all of its child tasks.
+/// You need to handle these considerations manually with a detached task.
```
good to explain what people are losing here! 👍

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -375,38 +392,26 @@ extension Task {
+/// A detached task runs to completion
+/// unless it is explicitly canceled by calling the `Task.Handle.cancel()` method.
+/// Specifically, dropping a detached task's handle
+/// doesn't cancel that task.
```
This wording is tricky; the task _always_ runs to completion, but the cancellation may cause it to return early or throw etc.

But since cancellation is cooperative, this means that if the task never checks cancellation, it'd finish as if nothing ever cancelled it (this is by design).

So we should not say that cancellation prevents running to completion, we can rather say that it impacts the task and how it will return.

ktoso replied:
Another note: cancellation does not prevent the task from being scheduled either, so it really is just informational to the task

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -430,36 +435,26 @@ public func detach<T>(
-/// A detached task always runs to completion unless it is explicitly cancelled.
-/// Specifically, dropping a detached tasks `Task.Handle` does _not_ automatically
-/// cancel given task.
+/// If the operation throws an error, this method rethrows that error.
```
This is quite not true, `detach` is not throwing.

What happens is that if the operation throws, then the returned Task.Handle's `try await get()` will throw. This is "kind of" like rethrowing but I don't think we can use that word here since "rethrows" is a specific thing in Swift.

amartini51 replied:
I'm changing "rethrows" to "propagates", to match what we call this in TSPL where the error makes its way back through a series of calls to throwing functions.

https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html#ID510

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -430,36 +435,26 @@ public func detach<T>(
+/// A detached task runs to completion
+/// unless it is explicitly canceled by calling the `Task.Handle.cancel()` method.
+/// Specifically, dropping a detached task's handle
+/// doesn't cancel that task.
```
same concern on wording as above

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -496,10 +491,13 @@ func _runAsyncHandler(operation: @escaping () async -> ()) {
-  /// Suspends the current task for _at least_ the given duration
-  /// in nanoseconds.
+  /// Suspends the current task,
+  /// and wait for at least the given duration before resuming it.
```
tenses are mixed up? `Suspends + waits` or `suspend + wait`?

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -496,10 +491,13 @@ func _runAsyncHandler(operation: @escaping () async -> ()) {
-  /// Suspends the current task for _at least_ the given duration
-  /// in nanoseconds.
+  /// Suspends the current task,
+  /// and wait for at least the given duration before resuming it.
```
can we bring back the emphasis on "_at least_"? It's important to highlight that it's not an exact sleep time guarantee

amartini51 replied:
I need to check editorial guidelines, but I think we avoid emphasis in abstracts because they show up on their own in places like navigation and Xcode quick help.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -522,12 +520,18 @@ extension Task {
+  /// Note that if this task is the highest-priority task in the system,
+  /// the executor immediately resumes execution of the same task.
+  /// As such,
+  /// this method isn't necessarily a way to avoid resource starvation.
```
(a bit weird formatting in this comment?)

amartini51 replied:
What looks weird to you?  If it's just the placement of line breaks, that's a convention we've used elsewhere in documentation to reduce noise in diffs.  Instead of breaking at the 80 column mark and rewrapping, we break lines at clause and sentence boundaries, so they don't need to be rewrapped after revising.

ktoso replied:
I see, that's a new technique to me but sounds fine I guess , thanks for explaining

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -566,22 +570,31 @@ extension Task {
+/// the unsafe task handle passed to the closure is always be non-nil
+/// because an asynchronous function always runs in the context of a task.
+/// However if you call this function from the body of a synchronous function,
+/// the unsafe task handle is `nil`
```
```suggestion
/// the unsafe task handle may be `nil`
``` 

feels more right somehow...? maybe not?

amartini51 replied:
If I call this method from sync code, and my backtrace was all sync back up the call stack, when would this return a non-nil value?

ktoso replied:
it would be nil then -- if the entire callstack is outside of swift concurrency APIs

amartini51 replied:
Thanks. If it's always `nil` in that case, I'll avoid writing "may be":

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -596,14 +609,22 @@ public func withUnsafeCurrentTask<T>(body: (UnsafeCurrentTask?) throws -> T) ret
-/// An `UnsafeCurrentTask` should not be stored for "later" access.
+/// To get an instance of `UnsafeCurrentTask` for the current task,
+/// call the `withUnsafeCurrentTask(body:)` method.
+/// Don't try to store an unsafe task handle
```
maybe just 
```suggestion
/// Don't store an unsafe task handle
```

?

amartini51 replied:
Maybe this phrasing doesn't work as well in reference here.  In TSPL, we use "don't try to do X" to describe something that won't work.  Here, you think you can store an unsafe task handle, but you really can't store it.  I think the counterpoint to my argument is that you *can* store it — just that storing the handle doesn't let you do anything useful.

ktoso replied:
storing it is actively unsafe and would lead to undefined behavior, so the stronger wording would be good hm

amartini51 replied:
Removed "try to" in commit 72322eb28db.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -596,14 +609,22 @@ public func withUnsafeCurrentTask<T>(body: (UnsafeCurrentTask?) throws -> T) ret
-/// The sub-set of APIs of `UnsafeCurrentTask` which also exist on `Task` are
+/// The subset of APIs of `UnsafeCurrentTask` which also exist on `Task` are
 /// generally safe to be invoked from any task/thread.
+/// ◊TR: Which ones, exactly, is that?  "Generally" isn't good enough here.
```
we can remove the "generally" the trick is "if an API exists on Task, it is also safe on UnsafeCurrentTask from other contexts; If it only exists on UnsafeCurrentTask, it is unsafe to call it from other tasks"

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -618,26 +639,26 @@ public struct UnsafeCurrentTask {
-  /// Operations on `Task` (unlike `UnsafeCurrentTask`) are safe to be called
-  /// from any other task (or thread).
+  /// The current task,
+  /// represented in a way that's safe to store for later use.
```
I much preferred the wording in terms of the operations that are safe to be called from any task, without necessarily implying storage.

Could we revive that?

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -618,26 +639,26 @@ public struct UnsafeCurrentTask {
   }
 
-  /// Returns `true` if the task is cancelled, and should stop executing.
+  /// A Boolean value that indicates whether the current task was canceled.
```
Do we really spell this as "Boolean"? The swift type is `Bool` 🤔 Or as the word I'd expect it as "boolean" 🤔

ktoso replied:
Maybe worth documenting that once this returns true it will always return true?

amartini51 replied:
Yes, this is the correct spelling per Apple Style Guide:

> Boolean (adj.) Note capitalization.

<https://help.apple.com/applestyleguide/#/apsg1a3a0436>

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -618,26 +639,26 @@ public struct UnsafeCurrentTask {
+  /// If there is no current task,
+  /// the value of this property is `Priority.default`.
+  /// ◊TR: When would this happen,
+  /// ◊TR: rather than withUnsafeCurrentTask() passing nil as its closure's argument?
```
Thanks for spotting this -- this will always return a value; the `static var currentPriority` needs the "if there is no current task" but the instance one does not.

ktoso commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -49,10 +49,10 @@ public func withTaskCancellationHandler<T>(
   ///
-  /// If no current `Task` is available, returns `false`, as outside of a task
-  /// context no task cancellation may be observed.
+  /// If there is no current task, the value of this property is `false`.
```
👍

ktoso commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -61,25 +61,16 @@ extension Task {
   }
 
-  /// Returns `true` if the task is cancelled, and should stop executing.
+  /// A Boolean value that indicates whether the task should stop executing.
```
I feel a bit uneasy dropping the "is cancelled" even though it reads like "cancelled is cancelled" i guess...

I somehow would like to indicate that all callers of isCancelled on this task, i.e. static funcs `Task.isCancelled` running within the task also will notice this.

amartini51 replied:
I was rewriting to avoid the content-free abstract problem of just repeating in English what the symbols in code already said.  Does the fact about all callers noticing cancellation belong here, or in the higher level discussion of cancellation at the top in that subsection?  Or maybe adding something here like:

> This value is used by other APIs, like the static `Task.isCancelled` property.

ktoso commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -61,25 +61,16 @@ extension Task {
-  ///
-  /// ### Suspension
-  /// This function returns instantly and will never suspend.
+  /// The error is always an instance of `Task.CancellationError`.
```
👍

ktoso commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -96,10 +87,11 @@ extension Task {
-  /// if the current task has been cancelled.
+  /// The `Task.checkCancellation()` method throws this error
+  /// if the current task has been canceled.
+  /// You can use this error in your own cancellation-checking code also.
```
If we wanted to we could document that it is allowed to throw other errors upon cancellation, but by convention returning this one is generally fine unless there are good reasons to use some other error.

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -33,52 +33,45 @@ extension Task {
+/// The tasks are canceled only if `cancelAll()` was invoked before returning,
+/// if the groups' task was canceled,
+/// or if the group body throws an error.
+/// ◊TR: Is it possible to throw an error here?
```
so there is `withThrowingTaskGroup` and `withTaskGroup` - throwing is only possible in the throwing one; perhaps the comment is currently copied (wrongly) to both?

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -33,52 +33,45 @@ extension Task {
-///       // some accumulation logic (e.g. sum += result)
+///         // some accumulation logic (e.g. sum += result)
 ///      }
+/// ◊TR: Which of the above do we prefer?  Why give two recommendations?
```
the group.next() can be useful if you want to specifically pick a few tasks, say "pick first two tasks, then cancel the rest"

```
let firstCompleted = await group.next() ?? 0
let secondCompleted = await group.next() ?? 0
group.cancelAll() // cancel the rest
return firstCompleted + secondCompleted
```

while the `for await` is sugar for "keep collecting until next returns nil" (so until we drained all tasks)

ktoso replied:
When i worked on those docs first we did not have the `for await` yet so I only "added" it later.

Perhaps we should do the following here:
- show the `for await` version and recommend it
- "if you need more control i.e. one-by-one picking completed tasks from the group you can use group.next()" and give a `guard let first = await group.next() else { ... }` example?

That should explain when to use that I hope :)

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -33,52 +33,45 @@ extension Task {
+/// The tasks must handle cancellation in some other way,
+/// such as returning the work completed so far, or returning `nil`.
+/// For tasks that need to handle cancellation by throwing an error,
+/// use the `withThrowingTaskGroup(of:returning:body:)` method instead.
```
nice thanks!

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -33,52 +33,45 @@ extension Task {
-/// If throwing is a good option for the kinds of tasks spawned by the group,
-/// consider using the `withThrowingTaskGroup` function instead.
+/// Canceling the task in which the group is running
+/// also cancels the group and all of its child tasks.
```
Let's also add this sneaky implication of this:

```swift
/// Therefore it is possible that a group.spawn() creates a task that is immediately cancelled
/// the moment it is created. This allows tasks to quickly return even as they are spawned from
/// an already cancelled group.
///
/// Alternatively, one might use `group.spawnUnlessCancelled` which does not spawn the tasks
/// if the group already is cancelled at the point of spawning. 
///
/// Using those functions one has full control over how to react to cancellation within a group:
/// some tasks may need to be run no-matter-what and some are better off not even spawned 
/// if they were to immediately notice the cancellation and not produce useful results.
```

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -104,60 +97,54 @@ public func withTaskGroup<ChildTaskResult: Sendable, GroupResult>(
+/// Even if their results are not collected explicitly and such task throws,
+/// and was not yet canceled,
+/// it may result in the `withTaskGroup` throwing.
+/// ◊TR: What does the above "such task throws" mean?
```
Let's remove that wording; we're not doing it.

It used to be about:

```
try await withThrowingTaskGroup(...) { g in
g.spawn { throw Boom() } 
}
```

so we never "collected" that boom; in practice we will ignore that error; We could have also decided to re-throw it, but that's not what we do nowadays

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -104,60 +97,54 @@ public func withTaskGroup<ChildTaskResult: Sendable, GroupResult>(
-/// error be rethrown by the group.
+/// If an error is thrown by one of the tasks in a task group,
+/// all of its remaining tasks are canceled,
+/// and the `withTaskGroup` method rethrows that error.
```
That simplification sadly is wrong, we must talk about the `try group.next()` throwing.

In this case:

```swift
{ group in 
group.spawn { throw Boom() } 
}
```

nothing is cancelled, while in this case all other tasks _are_ cancelled:

```swift
{ group in 
group.spawn { throw Boom() } 
try group.next()! // throws boom, exits the group, causes cancelAll
}
```

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -104,60 +97,54 @@ public func withTaskGroup<ChildTaskResult: Sendable, GroupResult>(
-/// Individual tasks throwing results in their corresponding `try group.next()`
-/// call throwing, giving a chance to handle individual errors or letting the
-/// error be rethrown by the group.
+/// If an error is thrown by one of the tasks in a task group,
```
we lost the section title about `Thrown errors` -- i think it's important to separate it from cancellation, can we revive it?

amartini51 replied:
I'm having trouble parsing that bit:

```
/// ### Thrown errors
 /// When tasks are added to the group using the `group.spawn` function, they may
 /// immediately begin executing. Even if their results are not collected explicitly
 /// and such task throws, and was not yet cancelled, it may result in the `withTaskGroup`
 /// throwing.
```

I think this is talking about how something like:

1. I spawn a child task
2. The child task might start executing immediately
3. The child task throws an error
4. Even though I didn't call `next()` the error propagates out of the whole `withTaskGroup` block

Is that right?

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -104,60 +97,54 @@ public func withTaskGroup<ChildTaskResult: Sendable, GroupResult>(
-///   - once the `withTaskGroup` returns the group is guaranteed to be empty.
-/// - if the body throws:
-///   - all tasks remaining in the group will be automatically cancelled.
+/// After this method returns, the task group is guaranteed to be empty.
```
👍 

Maybe worth a `### Postconditions`?

It kind of sneaks into the cancellation section now which isn't quite right

amartini51 replied:
Good catch that this doesn't belong under the cancellation section.  I'd like to avoid such a short section, so I've moved this up to be before the examples above.

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -220,21 +215,15 @@ public struct TaskGroup<ChildTaskResult: Sendable> {
-  ///
-  /// The `add` function will never (re-)throw errors from the `operation`.
-  /// Instead, the corresponding `next()` call will throw the error when necessary.
+  /// Adds a child task to the group.
```
Maybe 

```suggestion
  /// Spawns a child task in the group.
```

? We intend to familiarize developers with the "spawn" word

amartini51 replied:
Per editorial guidelines, I'm avoiding just repeating the same verb from a method name in its abstract when possible.  Using a different English word to describe what happens provides more information.

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -192,9 +179,17 @@ public func withThrowingTaskGroup<ChildTaskResult: Sendable, GroupResult>(
+/// To create a task group,
+/// call the `withTaskGroup(of:returning:body:)` method.
+///
+/// A task group most be used only within the task where it was created.
```
typo

```suggestion
/// A task group must be used only within the task where it was created.
```

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -220,21 +215,15 @@ public struct TaskGroup<ChildTaskResult: Sendable> {
-  ///     `false` otherwise (e.g. because the group `isCancelled`)
+  ///   - overridingPriority: The priority of the operation task.
+  ///     Omit this parameter or pass `.unspecified`
+  ///     to set the child task's priority to the priority of the group.
```
👍

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -220,21 +215,15 @@ public struct TaskGroup<ChildTaskResult: Sendable> {
+  ///     to set the child task's priority to the priority of the group.
+  ///   - operation: The operation to execute as part of the task group.
+  /// - Returns: `true` if the operation was added to the group successfully;
+  ///   otherwise; `false`.
```
Sorry this is my bad -- need to remove the `returns` -- the `spawn` does not return anything now.

Only `spawnUnlessCancelled` does return `bool`

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -260,21 +249,15 @@ public struct TaskGroup<ChildTaskResult: Sendable> {
+  ///     to set the child task's priority to the priority of the group.
+  ///   - operation: The operation to execute as part of the task group.
+  /// - Returns: `true` if the operation was added to the group successfully;
+  ///   otherwise; `false`.
```
👍

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -335,31 +329,7 @@ public struct TaskGroup<ChildTaskResult: Sendable> {
   ///         collected += value
   ///     }
   ///
```
we lost the thread-safety section which is worrying? We must explain that this should not be called from other tasks, correctness relies on it.

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -372,11 +342,13 @@ public struct TaskGroup<ChildTaskResult: Sendable> {
+  /// the task group is always empty.
+  /// It is guaranteed to be empty when returning from that body,
+  /// either because all child tasks have completed
+  /// or because they've been canceled.
```
```suggestion
  /// .
```

today's impl always forcefully awaits and that's correct

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -431,10 +402,18 @@ public struct TaskGroup<ChildTaskResult: Sendable> {
 
-/// A task group serves as storage for dynamically spawned, potentially throwing,
-/// child tasks.
+/// A task group serves as storage for dynamically spawned tasks,
```
```suggestion
/// A task group serves as storage for dynamically spawned,
```

"tasks" appears later in the sequence

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -472,21 +451,18 @@ public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
+  ///     Omit this parameter or pass `.unspecified`
+  ///     to set the child task's priority to the priority of the group.
+  ///   - operation: The operation to execute as part of the task group.
+  /// - Returns: `true` if the operation was added to the group successfully;
```
Sorry again here, need to remove the `returns`

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -513,21 +489,18 @@ public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
-  /// The `add` function will never (re-)throw errors from the `operation`.
-  /// Instead, the corresponding `next()` call will throw the error when necessary.
+  /// This method doesn't throw an error, even if the child task throws.
+  /// Instead, the next call to `TaskGroup.next()` rethrows that error.
```
The "corresponding" was important; it is not the "next `next()` call", consider this:

```swift
{ g in 
group.spawn { sleep(10); return 111 }
group.spawn { throw Boom() }

try await group.next() // boom!
}
 ```

since "next" is in completion order; so each next completion corresponds to some spawn, but the order is "by completion order" rather than "by submission order"

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -560,13 +533,24 @@ public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
+  ///     group.spawn { 1 }
+  ///     group.spawn { 2 }
+  ///
+  ///     print(await group.next())
```
I think our style since evolved to be `await print(group.next())`

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -588,35 +572,70 @@ public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
+  /// If the next child task throws an error
+  /// and you propagate that error from this method
+  /// out of the body of a `TaskGroup.withThrowingTaskGroup(of:returning:body:)` call,
+  /// then all remaining child tasks in that group are implicitly canceled.
```
yes right

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -588,35 +572,70 @@ public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
-  /// ### Thread-safety
-  /// Please note that the `group` object MUST NOT escape into another task.
-  /// The `group.next()` MUST be awaited from the task that had originally
-  /// created the group. It is not allowed to escape the group reference.
```
worried about losing thread safety section

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -588,35 +572,70 @@ public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
   ///
-  ///     print(await group.next())
-  ///     /// Prints "1" OR "2"
+  ///     print(await group.nextResult())
```
```suggestion
  ///     await print(group.nextResult())
```

minor

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -588,35 +572,70 @@ public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
+  ///
+  ///     for await try value in group {
+  ///         collected += value
+  ///     }
```
most of this comment is about `next()` now about `nextResult()`; let's only do a small section on:

```
guard let result = await group.nextResult() else {
  return // no task to wait on
}

switch result { 
case .success(let value): print(value)
case .failure(let error): print("oh no: \(error)")
}
```

maybe?

People will be using the `for await` and `next()` mostly

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -681,14 +701,17 @@ extension TaskGroup: AsyncSequence {
+  /// This iterator terminates after all tasks have completed successfully,
+  /// or after any task completes by throwing an error.
+  /// If a task completes by throwing an error,
+  /// no further task results are returned.
```
This wording is tricky.

This is legal:

```
group.spawn { 1 }
group.spawn { throw boom }
group.spawn { 2 }

do { 
  for await t in group { 
    print(t)
  }
} catch { just ignore the error } 

// wait again on tasks
for await t in group { 
  print(t)
}
```

for this snippet it is legal to return: 1 (1st loop), error, 2 (2nd loop); there's numerous combinations in which this may complete but ths one is legal too.

So it's important to explain that the _iterator made by the AsyncSequence_ will stop returning after something throws -- thats what makes the async loop and manual use of the iterator equivalent. But one is allowed to try waiting again

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -734,15 +766,17 @@ extension ThrowingTaskGroup: AsyncSequence {
-  /// are returned in *completion order*.
+  /// The elements returned by this iterator
+  /// appear in the order that the tasks *completed*,
+  /// not in the order that those tasks were added to the task group.
```
well worded! Can we add the same to the next() and nextResult() on the group itself?

amartini51 replied:
Where else did you want to add this?  I've checked the `next` and `nextResult` methods, and I think it's already on all of them.

ktoso commented:
Great first cut :) Some comments in line

amartini51 committed:
Drop trailing whitespace; add missing period.

amartini51 committed:
Incorporate tech review from @ktoso.

amartini51 commented:
@ktoso Thanks for the review.  I've made most of these changes in commit 2f12b74b14f but I need to come back tomorrow to pick up a couple fixes.  I also need to track the 'main' branch again and resolve merge conflicts.

amartini51 committed:
Restore thread-safety note.

amartini51 committed:
Current implementation waits, per @ktoso.

amartini51 committed:
Focus on nextResult() and show an example.

Folding the example @ktoso suggested into the existing code, and
removing content that was repeated here from next() but isn't relevant.

amartini51 committed:
Show how for-await interacts with errors.

amartini51 committed:
Remove tech review queries that were answered.

amartini51 commented:
@ktoso I rebased these changes on top of the Swift 5.5 release branch, and changed the PR to target that branch.  If you want to avoid re-reading the diff, you can look at the commits from `3c8b4f8deb6` on instead.

amartini51 committed:
Back out an unintended change from rebasing.

This was removed in commit f748e8d974870be90ec4b5b912b206be565c8adb.
Found this error and verified it's the only change outside of doc
comments by running:

    git diff release/5.5 | grep -v '^+ *///\|- *///' | grep -v '^@@' | grep -v '^ '

amartini51 committed:
Remove the second copy of a query.

amartini51 committed:
Track 'release/5.5' to resolve merge conflicts.

amartini51 committed:
Revise doc comments for changes from the release branch.

ktoso commented:
Oh sorry I missed the updates! Checking now :) Please ping me on chat if I miss something for a few days :)

amartini51 committed:
Incorporate editorial feedback.

For rdar://77330408

amartini51 commented:
@swift-ci please smoke test

amartini51 committed:
Track 'release/5.5' to resolve merge conflicts.

Conflicts in stdlib/public/Concurrency/TaskGroup.swift from async/spawn
renaming.  Resolved by copying the docs to the new symbols from their
old names.

ktoso reviewed (approved):
Looks good, thanks @amartini51 !

amartini51 committed:
Make wording a bit stronger.

amartini51 committed:
Remove tech review markers.

amartini51 committed:
Track 'release/5.5' to resolve merge conflicts.

Conflicts were just @availability marker changes.

amartini51 commented:
@swift-ci please smoke test

amartini51 commented:
@swift-ci Please test and merge.
